### PR TITLE
[WIP] Modal

### PR DESCRIPTION
**Pull request by ShadyXV**

### CODE REVIEW

**Comment by ShadyXV** [comment]
@syaau 🙂

File: src/Modal.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        return <Modal key={key} />
+      })}
+    </ModalsContainerComponent>
+  </ModalContext.Provider>
+};
+
+export const useModal = (Modal: ModalType) => {
+  const [visible, setVisible] = useState<boolean>(false);
+  const key = getUniqueId();
+  const context = useContext(ModalContext);
```
Comment (syaau) [inline src/Modal.tsx]: Use this as a controller `const controller = useContext(ModalContext);`

File: src/Modal.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  const hide = () => setVisible(false);
+
+  useEffect(() => {
+    if(visible) {
+      context.showModal(key, Modal);
+    } else {
+      context.hideModal(key);
+    }
+    return () => context.hideModal(key);
+  }, [visible]);
```
Comment (syaau) [inline src/Modal.tsx]: useEffect is to make sure that the modal is hidden when the parent component is unmounted. 
```
useEffect(() => modal.hide, []);
```

File: src/Modal.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        const Modal = modals[key];
+        return <Modal key={key} />
+      })}
+    </ModalsContainerComponent>
+  </ModalContext.Provider>
+};
+
+export const useModal = (Modal: ModalType) => {
+  const [visible, setVisible] = useState<boolean>(false);
+  const key = getUniqueId();
```
Comment (syaau) [inline src/Modal.tsx]: We are not going to maintain state here, as that would mean re-render in the components that are using modals. Instead, just create unique modal instance using the controller.
```typescript
const modal = useRef();
if (!modal.current) {
  modal.current = controller.createModal(Modal); // This function should generate a modal operator with a unique id
}

...


return modal;
```

**Pull request merged**